libs/ui/src/hooks/useSmartcard.test.tsx

```diff
@@ -0,0 +1,398 @@
+})
+
+test('with card reader and a pollworker card', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard(makePollWorkerCard(electionSampleDefinition.electionHash))
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('pollworker')
+})
+
+test('with card reader and an admin card', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard(makeAdminCard(electionSampleDefinition.electionHash))
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('admin')
+})
+
+test('with card reader and a gibberish card', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard('not JSON')
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('no card')
+})
+
+test('writing short value succeeds', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard()
+
+  const TestComponent = () => {
+    const [smartcard, hasCardReader] = useSmartcard({ card, hardware })
+    const [error, setError] = useState<Error>()
+
+    useEffect(() => {
+      void (async () => {
+        const result = await smartcard?.writeShortValue(
+          JSON.stringify(makeVoterCard(electionSample))
+        )
+        setError(result?.err())
+      })()
+    }, [!smartcard])
+
+    return (
+      <div>
+        {error?.message}{' '}
+        {hasCardReader ? smartcard?.data?.t ?? 'no card' : 'no card reader'}
+      </div>
+    )
+  }
+
+  render(<TestComponent />)
+
+  // trigger & complete write
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+
+  // update status
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+
+  screen.getByText('voter')
+})
+
+test('writing short value fails', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard()
+
+  const TestComponent = () => {
+    const [smartcard] = useSmartcard({ card, hardware })
+    const [error, setError] = useState<Error>()
+
+    useEffect(() => {
+      void (async () => {
+        const result = await smartcard?.writeShortValue(
+          JSON.stringify(makeVoterCard(electionSample))
+        )
+        setError(result?.err())
+      })()
+    }, [!smartcard])
+
+    return <div>{error?.message}</div>
+  }
+
+  jest.spyOn(card, 'writeShortValue').mockRejectedValue(new Error('oh no'))
+  render(<TestComponent />)
+
```

Comment:
FWIW, I find these comments above `advanceTimersAndPromises` to be helpful. More below if consistency is desirable.
```suggestion

  // trigger & complete write
```